jupyter_collaboration/handlers.py

```diff
@@ -385,18 +394,15 @@ class YDocRoomIdHandler(APIHandler):
     @web.authenticated
     @authorized
     async def put(self, path):
-        body = json.loads(self.request.body)
-        ws_url = f"{body['format']}:{body['type']}:"
-
         file_id_manager = self.settings["file_id_manager"]
 
         idx = file_id_manager.get_id(path)
         if idx is not None:
             # index already exists
             self.set_status(200)
-            ws_url += str(idx)
-            self.log.info("Request for Y document '%s' with room ID: %s", path, ws_url)
-            return self.finish(ws_url)
+            self.log.info("Request for Y document '%s' with room ID: %s", path, idx)
+            data = json.dumps({"file_id": idx, "session": DOCUMENT_SESSION})
```

Comment:
Why did you remove `format:type:` from `file_id`?

Reply:
Because we are not using it.

The WebSocketProvider requests the file id and sends the format and type in that request, then the backend retrieves the file id, composes the room id with `format:type:file_id`, and returns the room id. We have that info in the front. We only need the file id. We cannot even retrieve the file id when we don't have the format and type. So I don't think sending that info just to compose the room id in the backend makes sense. I believe it is better to retrieve the file id and compose the room id on the client side

Reply:
Now I'm thinking that maybe is better to use a get request since we only need to retrieve the file id.

Reply:
The `PUT /api/yjs/roomid/{path}` endpoint is for getting a room ID for a document path, format and type. What is returned is an implementation detail, it could also be a pure UUID. If it returns only the file ID, we cannot distinguish between a document e.g. as a plain file or as a notebook type.

Reply:
> Now I'm thinking that maybe is better to use a get request since we only need to retrieve the file id.

This has been discussed and we decided to use ` PUT`.

Reply:
Ooh, I got it. That makes sense.

Reply:
Let's not mix everything, `PUT /api/yjs/roomid/{path}` is for getting a room ID, as its name says.
Shared model version may be implemented using WebSocket subprotocols. Store version is independent.
BTW right now your implementation looks more like a "server session" than a "document session", since it's global. But it should be document-based, because updates in a YStore can be squashed on a document-based basis, right?

Reply:
> BTW right now your implementation looks more like a "server session" than a "document session", since it's global. But it should be document-based, because updates in a YStore can be squashed on a document-based basis, right?

This implementation is a server session to patch, as soon as possible, the issue where the content gets duplicated.

This implementation should be revised for JupyterLab v4. We should design a robust architecture with room sessions that contemplate squashing or removing the YStore, and checks the shared model's version.

> Let's not mix everything, `PUT /api/yjs/roomid/{path}` is for getting a room ID, as its name says.

In my opinion, everything is related. Maybe we should change this endpoint and create a more generic one to fetch room info.

When a client wants to join a collaborative session (a DocumentRoom), this client should meet some requirements.
* a compatible shared model
* non-outdated updates (after a squash or deletion of the store)

We could have different clients connecting to the same room (JupyterLab, Notebook, JupyterLite, etc.), and we should check those requirements before establishing a connection.

We could create a room session so every client that supports the requirement of that room can join. Using a more generic endpoint, we can fetch those requirements simultaneously (all of them are related and necessary to connect to a room). In addition, we need to check the requirement every time a client tries to join a collaborative session. As we see with the content duplication issue, the user may have old tabs open, and they will try to reconnect immediately. Anything could happen between a reconnection, even a new store version.

Reply:
All that is true, but I'm not sure we should change `PUT /api/yjs/roomid/{path}`. We could, however, accept the Y-WebSocket connection if all the requirements are met.

Reply:
And how do we request the requirement? Should we create a new endpoint?

Reply:
Maybe using WebSocket subprotocols for the Y-WebSocket? The client sends everything it supports, and the server checks them and accepts or closes the connection.

Reply:
> The client will unlikely send `everything` it supports because it will need to come with a specific version of `jupyter_ydoc` (whatever the language) and any doc provider. And I don't see us going into the trouble of supporting multiple versions within the same package. So my guess is that this will be simplified to _the client sends the "version" it supports_.

Reply:
I'm not sure supporting multiple versions is such an issue. The problem I see with the client sending only one supported version is that it could quickly lead to incompatible clients. But I'm not sure either this will be the case.

Reply:
> Maybe using WebSocket subprotocols for the Y-WebSocket? The client sends everything it supports, and the server checks them and accepts or closes the connection.

I don't think using subprotocols is the solution. In my opinion, we should validate the requirements before opening a connection.

After investigating a little bit, I believe the only way we can clean and reset a client, it is by disposing the shared model (destroying the YDoc and YWebSocketProvider) and instantiating a new one. I don't think it is possible to remove the updates from a YDoc and reinitialize it again (It doesn't make sense either).

The document session requirements are the information we need before establishing the communication. Once clients start communicating, those requirements can not change. Even in the case of squashing updates in the Store, we can only do it when the room is empty. Otherwise, we will need to:
1 - communicate to every client that there was a squash in the document, then on each client
2 - we will need to close the connection
3 - dispose the resources
4 - instantiate a new shared document
5 - open a new connection
That's inconvenient from a user perspective. Imagine you are working on a document, and every X minutes, the document closes (or asks you to close it) and reopens.

I also thought using subprotocols was the best solution, but after debugging this issue and investigating a little, I don't think it is the right solution.

Reply:
> I don't think using subprotocols is the solution. In my opinion, we should validate the requirements before opening a connection.

With WebSocket subprotocols, the requirements are validated before opening the connection.